python/examples/math/program.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+import asyncio
+import json
+from textwrap import dedent, indent
+from typing import TypeVar, Any, Callable, Awaitable, TypedDict, Annotated,  NotRequired, override, Sequence
+
+from typechat import (
+    Failure,
+    Result,
+    Success,
+    TypeChatModel,
+    TypeChatValidator,
+    TypeChatTranslator,
+    python_type_to_typescript_schema,
+)
+import collections.abc
+
+T = TypeVar("T", covariant=True)
+
+def Doc(s: str) -> str:
+    return s
+
+
+program_schema_text = '''
+// A program consists of a sequence of function calls that are evaluated in order.
+export type Program = {
+    "@steps": FunctionCall[];
+}
+
+// A function call specifies a function name and a list of argument expressions. Arguments may contain
+// nested function calls and result references.
+export type FunctionCall = {
+    // Name of the function
+    "@func": string;
+    // Arguments for the function, if any
+    "@args"?: Expression[];
+};
+
+// An expression is a JSON value, a function call, or a reference to the result of a preceding expression.
+export type Expression = JsonValue | FunctionCall | ResultReference;
+
+// A JSON value is a string, a number, a boolean, null, an object, or an array. Function calls and result
+// references can be nested in objects and arrays.
+export type JsonValue = string | number | boolean | null | { [x: string]: Expression } | Expression[];
+
+// A result reference represents the value of an expression from a preceding step.
+export type ResultReference = {
+    // Index of the previous expression in the "@steps" array
+    "@ref": number;
+};
+'''
+
+
+ResultReference = TypedDict(
+    "ResultReference", {"@ref": Annotated[int, Doc("Index of the previous expression in the 'steps' array")]}
+)
+
+FunctionCall = TypedDict(
+    "FunctionCall",
+    {
+        "@func": Annotated[str, Doc("Name of the function")],
+        "@args": NotRequired[Annotated[list["Expression"], Doc("Arguments for the function, if any")]],
+    },
+)
+
+JsonValue = str | int | float | bool | None | dict[str, "Expression"] | list["Expression"]
+Expression = JsonValue | FunctionCall | ResultReference
+
+JsonProgram = TypedDict("Program", {"@steps": list[FunctionCall]})
+
+
+async def evaluate_json_program(program: JsonProgram, onCall: Callable[[str, Sequence[Expression]], Awaitable[Expression]]) -> Expression | Sequence[Expression]:
+    results: list[Expression] | Expression = []
+
+    async def evaluate_array(array: Sequence[Expression]) -> Sequence[Expression]:
+        return await asyncio.gather(*[evaluate_call(e) for e in array]) # type: ignore
+
+    async def evaluate_object(expr: FunctionCall):
+        if "@ref" in expr:
+            index = expr["@ref"]
+            if index < len(results):
```

Comment:
You probably need to make sure `expr` is an `int`, and report otherwise, and report if it is out of bounds.

Reply:
Let's sync up on this. As I see it, the goal of the examples is to show devs how to use the library. In this case, the program APIs can be presented using the typeddict-of-callables approach (schema.py) or the protocol approach (schemav2.py).  We support both approaches and the devs are free to choose either one.  I'm going to add some Readme files that will explain how to switch between the different schema definitions